### DISCUSSION TITLE
v0.23.0-rc: how to wrap an entire hook chain in a single DB transaction?

### DISCUSSION THREAD

**Discussion by Sleepful** [pocketbase/pocketbase#5783] [Q&A] [answered]
Looking at [releases/tag/v0.23.0-rc](https://github.com/pocketbase/pocketbase/releases/tag/v0.23.0-rc), I can see:
> New hooks allowing better control over the execution chain and error handling (including wrapping an entire hook chain in a single DB transaction).

I want to know how to wrap the entire hook chain in a transaction.
I reviewed these documents:
- https://pocketbase.io/v023upgrade/go/#event-hooks-changes
- https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc8/core
But I could not find an answer.

I noticed that the hooks now look like this though:

```go
app.OnRecordUpdateRequest().BindFunc(func(e *core.RecordRequestEvent) error {
	// before update

	// I would do something with the database here and I would want it
	// to be in the same transaction as the update operation

    if err := e.Next(); err != nil {
        return err
    }
	// after update
    return nil
})
```
Thanks, very excited about v23! :)

**Comment by ganigeorgiev** [comment, marked as answer]
Reassigning `e.App` with the transactional one should be enough, something like:

```go
app.OnRecordUpdateRequest().BindFunc(func(e *core.RecordRequestEvent) error {
    originalApp := e.App
    txErr := e.App.RunInTransaction(func(txApp core.App) error {
        e.App = txApp

        // before update

        // I would do something with the database here and I would want it
        // to be in the same transaction as the update operation

        if err := e.Next(); err != nil {
            return err
        }

        // after update

        return nil
    })
    e.App = originalApp

    return txErr
})
```

With that said, avoid slow/blocking operations (sending emails, file upload to S3, etc.) as part of a read&write transactions as it requires db lock and may degrade your app performance.

**Reply by Sleepful** [reply]
Thanks for the quick answer Gani.

I am curious about this:

> avoid slow/blocking operations (sending emails, file upload to S3, etc.) as part of a read&write transactions as it requires db lock and may degrade your app performance

I was investigating about the SQLite transactions.
 
CMIIAW:

In a read transaction it is OK to have slow operations such as a file upload because the read transactions do not block the database. 

However, once the transaction is upgraded to a write transaction, the database locks any other writers until the transaction finishes. This would degrade performance significantly.

Is this correct?

> If a write statement occurs while a read transaction is active, then the read transaction is upgraded to a write transaction if possible. If some other database connection has already modified the database or is already in the process of modifying the database, then upgrading to a write transaction is not possible and the write statement will fail with [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy).

From: https://www.sqlite.org/lang_transaction.html

**Reply by Sleepful** [reply]
Second question: 

If the transaction shown in your sample code above fails with [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) due to another writer updating the DB first, will Pocketbase retry the operation?

It is a strange edge-case, but I believe it could happen if the transaction begins as a read-transaction, then a write occurs from a different connection, and then the transaction we are talking about tries to upgrade to a write-transaction, which would make it fail with `SQLITE_BUSY`

**Reply by ganigeorgiev** [reply]
PocketBase maintains 2 pools - 1 for read allowing 120 max connections and 1 for writes/transactions for only 1 connection.

Meaning that `SQLITE_BUSY` cannot occur from just a write/transaction because the queries will be queued on application level. When the transaction is mixed (read&write) and there is another read happening then it is possible busy or locked error to occur but for this we have the `busy_timeout` PRAGMA + additional auto/fail retry on application level (_this in practice means that other users will have to wait their turn, which is what I meant by "degrade your app performance"_).

In any way, the easiest way to rule out your concerns is to test it.

**Reply by Sleepful** [reply]
> 1 for writes/transactions for only 1 connection.

This one right?
https://github.com/pocketbase/pocketbase/blob/bc83fddaf2c35ef77b49e9fb7770d93ecbf3d749/core/db_tx.go#L14-L16

Configured here:
https://github.com/pocketbase/pocketbase/blob/bc83fddaf2c35ef77b49e9fb7770d93ecbf3d749/core/base.go#L1115

> When the transaction is mixed (read&write) and there is another read happening then it is possible busy or locked error to occur

So, if I understand correctly, the operations inside `RunInTransaction` will use the `NonconcurrentDB` which works as a queue for writers. For this reason any operations inside `RunInTransaction` will always succeed, and _not_ error with `SQLITE_BUSY`, because there will never be two concurrent writes.

The one issue that you mentioned is that a write connection (using the one connection available from `NonconcurrentDB`) could bottleneck and delay all other write operations, as every operation waits in a queue for the transaction to end.  

I believe even read-only transactions should be successful all the time because a read-only transaction might read a value that later gets updated, but it is not going to fail.

Does this sound correct? 

Just making sure I don't miss edge cases that could cause unhandled failures, which would be rather difficult to test.

Thank you.

**Reply by ganigeorgiev** [reply]
I think we are going in circles and will stop engage further. I don't know how to explain it better than what I've already written above.

My suggestion remains - avoid long (aka. slow/blocking) operations inside of a transaction.

**Reply by Sleepful** [reply]
Not circles, just clearing up the possibilities of SQLITE errors on the application level, which would be important for anyone worrying about concurrency. 

Especially considering that the PB documentation doesn't go into detail about this, and users cannot rely on knowledge about default SQLite behavior because of the application layer that you have written. 

I think I got all the relevant info from your comments + code though, so don't worry about the discussion any more.

Your suggestion is noted and makes sense, to avoid blocking write operations.
